pkg/controller/backup_controller.go

```diff
@@ -666,17 +805,100 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
 		return err
 	}
 
-	if errs := persistBackup(backup, backupFile, logFile, backupStore, c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)), volumeSnapshots, volumeSnapshotContents); len(errs) > 0 {
+	if errs := persistBackupBeforeUploadComplete(backup, backupFile, logFile, backupStore, volumeSnapshots, volumeSnapshotContents); len(errs) > 0 {
 		fatalErrs = append(fatalErrs, errs...)
 	}
 
-	c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup completed")
+	snapshotInProgress := false
+	if features.IsEnabled(velerov1api.UploadProgressFeatureFlag) {
+		snapshotInProgress = checkAndUpdateItemSnapshotsProgress(backup.ItemSnapshots, logger)
+	}
+
+	if snapshotInProgress {
+		// Prepare for upload progress monitoring
+		switch {
+		case len(fatalErrs) > 0:
+			// If we had fatal errors in the backup, we will just fail here and not monitor upload progress
+			// ItemSnapshotsAttempted/Completed may be incorrect
+			backup.Status.Phase = velerov1api.BackupPhaseFailed
+		case logCounter.GetCount(logrus.ErrorLevel) > 0:
+			backup.Status.Phase = velerov1api.BackupPhaseUploadingPartialFailure
+			c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup partially failed but snapshot uploads outstanding, will complete in background")
+		default:
+			backup.Status.Phase = velerov1api.BackupPhaseUploading
+			c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Snapshot uploads outstanding, will complete in background")
+		}
+	} else {
+		updateBackupStatusForSnapshotState(backup, c.logger)
+
+		switch {
+		case len(fatalErrs) > 0:
+			backup.Status.Phase = velerov1api.BackupPhaseFailed
+		case logCounter.GetCount(logrus.ErrorLevel) > 0:
+			backup.Status.Phase = velerov1api.BackupPhasePartiallyFailed
+		default:
+			backup.Status.Phase = velerov1api.BackupPhaseCompleted
+		}
+	}
 
+	if backup.Status.Phase != velerov1api.BackupPhaseUploading &&
+		backup.Status.Phase != velerov1api.BackupPhaseUploadingPartialFailure {
+		// We're done so set the CompletionTimestamp now
+		backup.Status.CompletionTimestamp = &metav1.Time{Time: c.clock.Now()}
+		recordBackupDurationMetric(backup.Backup, c.metrics)
+		if errs := persistBackupFinal(backup, backupStore); len(errs) > 0 {
+			fatalErrs = append(fatalErrs, errs...)
+		}
+		c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup completed")
+	} else {
+		c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup completed, snapshots uploading in the background")
+	}
 	// if we return a non-nil error, the calling function will update
```

Comment:
For some reason I can't comment on the question about BackupItemAction vs ItemSnapshotter and recursion/additional items.

What is behind ItemSnapshotter may be something that handles the snapshot of a tree of items.  For example, if you have a snapshot implemented by an Operator, it may handle a number of items behind the snapshot.  Some of these may be snapshotted, others may simply need to be ignored.  For example, if we have a database that produces a logical dump of itself we would want to ignore any PVs that it uses.  This is why SnapshotItem can return a list of other items that it has handled (HandledItems).  

There's currently no way to communicated between plugins so co-ordinating a BackupItemAction and an ItemSnapshotter will be difficult.

Note that this is already merged and is not part of this PR.  The design can be revisited, but that should be handled in a separate PR, not in comments on this one.

Reply:
As far as refactoring, I pulled the metrics update out in a separate function.  I don't think it makes sense to merge the metrics update and the update of the resource.